instance/tasks.py

```diff
@@ -339,3 +344,111 @@ def cleanup_old_betatest_users():
+    """
+
+    def __init__(self):
+        self.region: str = getattr(settings, "OPENSTACK_REGION", "gra")
+        self.recipient: list = [email for _, email in getattr(settings, "ADMINS", [])]
+        self.threshold: int = getattr(settings, "KILL_ZOMBIES_WARNING_THRESHOLD")
+
+    def send_email(self, subject: str, body: str) -> int:
+        """
+        Utility method for sending emails
+        """
+        if self.recipient:
+            email = EmailMessage(
+                subject, body, settings.DEFAULT_FROM_EMAIL, self.recipient
+            )
+            return email.send()
+        else:
+            logging.info("Email recipient is undefined. Skipping email.")
+            return None
+
+    def trigger_warning(self):
+        """
+        Warns when more than `threshold` VMs will be terminated
+        """
+        subject = "{} Zombie instances over current threshold ({})".format(
+            datetime.now().strftime("%Y-%m-d"), self.threshold
+        )
+        body = (
+            "The number of instances in region {} to be terminated"
+            "is over KILL_ZOMBIES_WARNING_THRESHOLD ({}).\n"
+        ).format(self.region, self.threshold)
+        logging.info(
+            "%s\nSending an email notification to: %s",
+            body, self.recipient
+        )
+        self.send_email(subject, body)
+
+    def handle_dry_run_output(self, stdout: str) -> int:
+        """
+        If there are servers to terminate, a `kill_zombies` dry run will warn:
+        "Would have terminated {} zombies" to stdout.
+        """
+        if "No servers found in region" in stdout:
+            return 0
+        try:
+            return int(stdout.split("Would have terminated")[1].split()[0])
+        except IndexError:
+            logger.info("Received unexpected input from dry run")
+            return None
+
+
+if getattr(settings, "KILL_ZOMBIES_ENABLED"):
+    @db_periodic_task(crontab(
+        minute=KILL_ZOMBIES_CRON_SCHEDULE[0],
+        hour=KILL_ZOMBIES_CRON_SCHEDULE[1],
+        day=KILL_ZOMBIES_CRON_SCHEDULE[2],
+        month=KILL_ZOMBIES_CRON_SCHEDULE[3],
+        day_of_week=KILL_ZOMBIES_CRON_SCHEDULE[4],
+    ))
+    def kill_zombies_periodically_task():
+        """
+        A wrapper for `kill_zombies_periodically` so it only
+        exists when KILL_ZOMBIES_ENABLED is true
+        """
+        kill_zombies_periodically()
+
+
+def kill_zombies_periodically():
+    """
+    Runs the 'kill_zombies' command to terminate servers
+    whose manual termination didn't succeed.
+    Defaults schedule is daily at 12AM
+    """
+    task_runner = KillZombiesRunner()
+    logging.info("Executing periodic task `kill_zombies_periodically`")
+    try:
+        dry_run_output, output = "", ""
+        dry_run_output = call_command(
+            "kill_zombies",
+            region=task_runner.region,
+            dry_run=True,
+        )
+        num_of_zombies = task_runner.handle_dry_run_output(dry_run_output)
```

Comment:
Since you have abstracted away all the logic in the `KillZombiesRunner` class, this function shouldn't have to bother about the "how" part and should just ask `task_runner.get_zombie_servers_count()` to give it the number? After all, it is the "runner" class according to its name.

Reply:
I'll apply this.